### PR TITLE
Dbot train classifier v3

### PR DESCRIPTION
**Pull request by yguezpa**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- In Progress

## Related Issues
fixes: 28937

## Description
Migrate "DBotTrainTextClassifierV2" to python3

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 1]
```diff
@@ -112,30 +114,37 @@ def get_data_with_mapped_label(data, labels_mapping, tag_field):
...
+    model_labels = PhishingModel.get_model_labels()
+    model_data = demisto_ml.phishing_model_dumps(PhishingModel)
 
     res = demisto.executeCommand('createMLModel', {'modelData': model_data,
                                                    'modelName': model_name,
                                                    'modelLabels': model_labels,
-                                                   'modelOverride': model_override,
-                                                   'modelExtraInfo': {'threshold': threshold}
+                                                   'modelOverride': model_override
+                                                   # 'modelExtraInfo': {'threshold': threshold}
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py:129]: you can remove the comment out, it's ok, just cast threshold to float

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 2]
```diff
@@ -216,12 +225,12 @@ def get_ml_model_evaluation(y_test, y_pred, target_accuracy, target_recall, deta
 
 def validate_data_and_labels(data, exist_labels_counter, labels_mapping, missing_labels_counter):
     labels_counter = Counter([x[DBOT_TAG_FIELD] for x in data])
-    labels_below_thresh = [label for label, count in labels_counter.items() if count < MIN_INCIDENTS_THRESHOLD]
+    labels_below_thresh = [x for x, count in labels_counter.items() if count < MIN_INCIDENTS_THRESHOLD]
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py:228]: why rename everything to x in the list expressions?

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 3]
```diff
@@ -112,30 +114,37 @@ def get_data_with_mapped_label(data, labels_mapping, tag_field):
...
                                                    })
     if is_error(res):
         return_error(get_error(res))
     confusion_matrix_no_all = {k: v for k, v in confusion_matrix.items() if k != 'All'}
     confusion_matrix_no_all = {k: {sub_k: sub_v for sub_k, sub_v in v.items() if sub_k != 'All'}
                                for k, v in confusion_matrix_no_all.items()}
+
+    y_test_pred_prob = [float(x) for x in y_test_pred_prob]
     res = demisto.executeCommand('evaluateMLModel',
                                  {'modelConfusionMatrix': confusion_matrix_no_all,
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py:139]: we need to change the confusion matrix to be confusion matrix with considering the threshold. ask @lperryp about it

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml [hunk 4]
```diff
@@ -150,29 +150,31 @@ commonfields:
...
 script: '-'
+subtype: python3
 system: false
 tags:
 - ml
 timeout: 12µs
 type: python
-subtype: python2
-dockerimage: demisto/dl:1.4
+dockerimage: demisto/mllocal:latest
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml:177]: I think we can create the docker image in the repo and assign it here

**Review by erezh31 — CHANGES_REQUESTED** [review]
small changes
@lperryp please review too

@yguezpa we can move to the next script of DBotPredictPhishingWords

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 5]
```diff
@@ -112,30 +114,37 @@ def get_data_with_mapped_label(data, labels_mapping, tag_field):
...
 
-def store_model_in_demisto(model_name, model_override, X, y, confusion_matrix, threshold, y_test_true, y_test_pred,
-                           y_test_pred_prob):
-    model = demisto_ml.train_text_classifier(X, y, True)
-    model_data = demisto_ml.encode_model(model)
-    model_labels = demisto_ml.get_model_labels(model)
+def store_model_in_demisto(model_name, model_override, X, y, confusion_matrix, threshold, tokenizer, y_test_true,
+                           y_test_pred,
+                           y_test_pred_prob, target_accuracy):
+    PhishingModel = demisto_ml.PhishingModel(tokenizer)
```
Comment (lperryp) [inline Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py:120]: I think it's better to change variable name to"`phishing_model`, might lead to errors with the imports in future

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py [hunk 6]
```diff
@@ -325,6 +335,8 @@ def get_X_and_y_from_data(data, text_field):
 
 
 def main():
+    tokenizer_script = demisto.args().get('tokenizerScript', None)
```
Comment (lperryp) [inline Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py:338]: what is `tokenizerScript`?
1. If it's aimed for the user to change, you need to expose it in the yml 
2. if same argument exists in other automations, try to be consistent with it

**Review by lperryp — CHANGES_REQUESTED** [review]

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py (hunk 6 above)
Comment (erezh31) [reply Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py:338]: @lperryp it's like the tokenizer method, we have only 1 now. 
It's suppose to be part of the script inputs, @yguezpa please add this with default value of "nlp", also you can change the name of the argument to be "tokernizerMethod"

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py (hunk 5 above)
Comment (erezh31) [reply Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py:120]: agree, name convention is lower case underscore

**Comment by yguezpa** [comment]
@erezh31 
Changes you and Lior asked
Change code and unitest for Predict automation

**Comment by yguezpa** [comment]
@erezh31 changed docker

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml [hunk 7]
```diff
@@ -140,7 +152,14 @@ args:
   isArray: false
   name: originalTextFields
   predefined:
-  - ''
+    - ''
+  required: false
+  secret: false
+- default: false
+  defaultValue: nlp
```
Comment (erezh31) [inline Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.yml:159]: @lperryp is the correct value with your script?

File: Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py (hunk 3 above)
Comment (erezh31) [reply Packs/Base/Scripts/DBotTrainTextClassifierV2/DBotTrainTextClassifierV2.py:139]: @yguezpa plz work with @lperryp to change the confusion matrix to be the one with the threshold considiration

**Review by erezh31 — CHANGES_REQUESTED** [review]

**Comment by lperryp** [comment]
@yguezpa to change the confusion matrix to be the one with the threshold consideration:
in `output_model_evaluation` change:
 `return confusion_matrix_no_thresh` to `return confusion_matrix_at_thresh`

**Comment by lperryp** [comment]
@erezh31 @yguezpa - before merging - does this change support current format of models?

**Comment by yguezpa** [comment]
@kirbles19 could you please review documentation ? Thanks !

**Comment by kirbles19** [comment]
@yguezpa Done.

**Pull request closed without merging**